JASP-Engine/JASP/R/mediationanalysis.R

```diff
@@ -0,0 +1,573 @@
+    tottab$addColumnInfo(name = "ci.upper", title = "Upper",      type = "number", format = "sf:4;dp:3",
+                         overtitle = paste0(options$ciWidth * 100, "% Confidence Interval"))
+    tottab$addFootnote(se_message)
+    pe_tot <- pe[pe$op == ":=" & substr(pe$lhs, 1, 3) == "tot",]
+    
+    tottab[["lhs"]]      <- .unv(rep(options$predictor, length(options$dependent)))
+    tottab[["op"]]       <- rep("\u2192", nrow(pe_tot))
+    tottab[["rhs"]]      <- .unv(rep(options$dependent, each = length(options$predictor)))
+    tottab[["est"]]      <- pe_tot$est
+    tottab[["se"]]       <- pe_tot$se
+    tottab[["z"]]        <- pe_tot$z
+    tottab[["pvalue"]]   <- pe_tot$pvalue
+    tottab[["ci.lower"]] <- pe_tot$ci.lower
+    tottab[["ci.upper"]] <- pe_tot$ci.upper
+    
+  }
+
+
+  ## residual covariances
+  if (options[["showres"]] && length(c(options$mediators, options$dependent)) > 2) {
+    pecont[["res"]] <- restab <- createJaspTable(title = "Residual covariances")
+    restab$addColumnInfo(name = "lhs",      title = "",           type = "string")
+    restab$addColumnInfo(name = "op",       title = "",           type = "string")
+    restab$addColumnInfo(name = "rhs",      title = "",           type = "string")
+    restab$addColumnInfo(name = "est",      title = "Estimate",   type = "number", format = "sf:4;dp:3")
+    restab$addColumnInfo(name = "se",       title = "Std. Error", type = "number", format = "sf:4;dp:3")
+    restab$addColumnInfo(name = "z",        title = "z-value",    type = "number", format = "sf:4;dp:3")
+    restab$addColumnInfo(name = "pvalue",   title = "p",          type = "number", format = "dp:3;p:.001")
+    restab$addColumnInfo(name = "ci.lower", title = "Lower",      type = "number", format = "sf:4;dp:3",
+                         overtitle = paste0(options$ciWidth * 100, "% Confidence Interval"))
+    restab$addColumnInfo(name = "ci.upper", title = "Upper",      type = "number", format = "sf:4;dp:3",
+                         overtitle = paste0(options$ciWidth * 100, "% Confidence Interval"))
+    restab$addFootnote(se_message)
+    pe_res <- pe[pe$op == "~~" &
+                   pe$lhs != pe$rhs &
+                   !.unv(pe$lhs) %in% options$predictor &
+                   !.unv(pe$lhs) %in% options$confounds,]
+
+    restab[["lhs"]]      <- .unv(pe_res$lhs)
+    restab[["op"]]       <- rep("\u2194", nrow(pe_res))
+    restab[["rhs"]]      <- .unv(pe_res$rhs)
+    restab[["est"]]      <- pe_res$est
+    restab[["se"]]       <- pe_res$se
+    restab[["z"]]        <- pe_res$z
+    restab[["pvalue"]]   <- pe_res$pvalue
+    restab[["ci.lower"]] <- pe_res$ci.lower
+    restab[["ci.upper"]] <- pe_res$ci.upper
+  }
+
+}
+
+.medRsquared <- function(jaspResults, medResult, options, errors) {
+  if (!options$rsquared || !is.null(jaspResults[["rsquared"]])) return()
+  
+  jaspResults[["rsquared"]] <- tabr2 <- createJaspTable("R-Squared")
```

Comment:
It's better (this goes for all output elements) to only assign things to jaspResults once they have some markup and are actually presentable. So in this specific example if you assign ` jaspResults[["rsquared"]] <- tabr2` after the `addColumnInfo`'s you avoid jaspResults sending empty tables to the output (no columns, no row of dots). 

A visual example of why it is better not to assign empty elements without any markup to jaspResults:

<img width="930" alt="Screenshot 2019-07-05 at 10 09 45" src="https://user-images.githubusercontent.com/18737241/60708023-5f134380-9f0d-11e9-8098-1e96d8e907ea.png">

Reply:
I have moved the table assignment to after the `addColumnInfo`, but I don't think this solves the specific issue you refer to.

Reply:
Yeah the screenshot was a more extreme, but unrelated example. That's caused by the fact that you compute the model before you create any of the tables.

The nice thing of the pass-by-reference mechanics of jaspResults is that you can just compute the model where you actually need it, and don't have to necessarily compute it first and then pass it around among the subfunctions. [An example of that can be found in this file](https://github.com/jasp-stats/jasp-desktop/blob/b23d700bbb79f395290643fb7002137fea75db29/JASP-Engine/JASP/R/regressionlinearbayesian.R). There you'll find that `.basregComputeModel()` is not called within the main function, and no model is passed around. Therefore you can already show output to the user before calculating the model.